### PR TITLE
Refactor/add dependency injection

### PR DESCRIPTION
**Pull request by mrsasha**
@skraynick and here's the pull request. good luck! ;) btw, the last is the merge commit because I first wanted to merge your sugar commits and maybe fix any problems, but merging went more or less smoothly, just some minor fixes.

one note: this branch breaks travis building again, it's complaining mostly about volley now. That will be fixed in another branch when I refactor volley out, but I wanted first to push this one out.

also, the bootstrapper is now integrated and not a separate library as before.

### CODE REVIEW

**Comment by skraynick** [comment]
@mrsasha awesome(ly) big. It will focus my attention on something other than sad stuff. I will read through over the next couple of days.  I don't have reliable internet at the moment, sorting it out.. so working on stuff from coffee shops mainly.  ;)

**Comment by mrsasha** [comment]
ah, the disconnected life! no hurry, everything's under control.

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/SensorbergTestApplication.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    public void onCreate() {
+        super.onCreate();
+        setComponent(buildComponentAndInject());
+        SensorbergApplicationBootstrapper bootstrapper = new SensorbergApplicationBootstrapper(this);
+    }
+
+    public Component buildComponentAndInject() {
+        return TestComponent.Initializer.init(this);
+    }
+
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/SensorbergTestApplication.java]: no space needed.

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/di/TestComponent.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    final class Initializer {
+
+        public static TestComponent init(Application app) {
+            return DaggerTestComponent.builder()
+                    .testProvidersModule(new TestProvidersModule(app))
+                    .build();
+        }
+    }
+
+}
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/di/TestComponent.java]: space needed.

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/di/TestComponent.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    @NoArgsConstructor(access = AccessLevel.PRIVATE)
+    final class Initializer {
+
+        public static TestComponent init(Application app) {
+            return DaggerTestComponent.builder()
+                    .testProvidersModule(new TestProvidersModule(app))
+                    .build();
+        }
+    }
+
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/di/TestComponent.java]: no space needed.

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/internal/http/OkVolleyShouldCacheTheSettingsWithEtags.java [hunk 4]
```diff
@@ -63,23 +73,22 @@ public void setUp() throws Exception {
         queue = new RequestQueue(new DiskBasedCache(cacheDir), network);
         queue.start();
 
-        when(testPlattform.getVolleyQueue()).thenReturn(queue);
-        tested = new OkHttpClientTransport(testPlattform, null);
+        tested = new OkHttpClientTransport(queue, clock, testPlatformIdentifier, true);
         tested.setApiToken(TestConstants.API_TOKEN);
         startWebserver();
     }
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/internal/http/OkVolleyShouldCacheTheSettingsWithEtags.java]: No space

Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/internal/http/OkVolleyShouldCacheTheSettingsWithEtags.java]: let's take out the extra space.

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/internal/http/OkVolleyShouldCacheTheSettingsWithEtags.java [hunk 5]
```diff
@@ -99,9 +108,9 @@ public void test_cache_revalidation_with_etag() throws Exception {
     public void test_cache_revalidation_with_header() throws Exception {
         enqueue(R.raw.response_etag_001, R.raw.response_etag_002);
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/internal/http/OkVolleyShouldCacheTheSettingsWithEtags.java]: above we have no space after enqueue, but here we do.  let's make consistent.

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/internal/http/TransportShould.java [hunk 6]
```diff
@@ -77,29 +90,32 @@ public void test_should_forward_the_layout_upload_interval_to_the_settings() thr
...
-            public void onFailure(VolleyError e) {
-                Log.d("FOO", "onFailure" + e.getLocalizedMessage());
-            }
 
-            @Override
-            public void onSettingsFound(JSONObject settings) {
-                fail();
-            }
-        });
+        //TODO
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/internal/http/TransportShould.java]: could we add a comment to what needs to be done...

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/scanner/TheBackgroundScannerShould.java [hunk 7]
```diff
@@ -1,102 +1,128 @@
...
+        testHandlerManager.getCustomClock().setNowInMillis(DefaultSettings.DEFAULT_BACKGROUND_SCAN_TIME + 1);
 
-        platform.clock.setNowInMillis(Settings.DEFAULT_BACKGROUND_SCAN_TIME + Settings.DEFAULT_BACKGROUND_WAIT_TIME);
-        platform.clock.setNowInMillis(Settings.DEFAULT_BACKGROUND_SCAN_TIME + Settings.DEFAULT_BACKGROUND_WAIT_TIME + 1);
+        testHandlerManager.getCustomClock().setNowInMillis(DefaultSettings.DEFAULT_BACKGROUND_SCAN_TIME + DefaultSettings.DEFAULT_BACKGROUND_WAIT_TIME);
+        testHandlerManager.getCustomClock()
+                .setNowInMillis(DefaultSettings.DEFAULT_BACKGROUND_SCAN_TIME + DefaultSettings.DEFAULT_BACKGROUND_WAIT_TIME + 1);
 
         ScannerListener mockListener = Mockito.mock(ScannerListener.class);
         tested.addScannerListener(mockListener);
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/scanner/TheBackgroundScannerShould.java]: Let's take out the space.

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/scanner/TheForegroundScannerShould.java [hunk 8]
```diff
@@ -102,24 +131,28 @@ public void test_background_scan_times_are_applied(){
...
 
         //finish the foreground wait time
-        platform.clock.setNowInMillis(Settings.DEFAULT_FOREGROUND_SCAN_TIME + Settings.DEFAULT_FOREGROUND_WAIT_TIME - 1);
-        platform.clock.setNowInMillis(Settings.DEFAULT_FOREGROUND_SCAN_TIME + Settings.DEFAULT_FOREGROUND_WAIT_TIME);
-        platform.clock.setNowInMillis(Settings.DEFAULT_FOREGROUND_SCAN_TIME + Settings.DEFAULT_FOREGROUND_WAIT_TIME + 1);
+        testHandlerManager.getCustomClock()
+                .setNowInMillis(DefaultSettings.DEFAULT_FOREGROUND_SCAN_TIME + DefaultSettings.DEFAULT_FOREGROUND_WAIT_TIME - 1);
+        testHandlerManager.getCustomClock().setNowInMillis(DefaultSettings.DEFAULT_FOREGROUND_SCAN_TIME + DefaultSettings.DEFAULT_FOREGROUND_WAIT_TIME);
+        testHandlerManager.getCustomClock()
+                .setNowInMillis(DefaultSettings.DEFAULT_FOREGROUND_SCAN_TIME + DefaultSettings.DEFAULT_FOREGROUND_WAIT_TIME + 1);
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/scanner/TheForegroundScannerShould.java]: let's delete the space.

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/scanner/TheForegroundScannerShould.java [hunk 9]
```diff
@@ -129,30 +162,35 @@ public void test_background_wait_starts(){
...
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/scanner/TheForegroundScannerShould.java]: no need for space.

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/testUtils/TestBluetoothPlatform.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    public static final BeaconId EXPECTED_BEACON_1 = new BeaconId(Utils.hexStringToByteArray(BEACON_ID_1));
+    public static final BeaconId EXPECTED_ALIEN_1 = new BeaconId(Utils.hexStringToByteArray(ALIEN_ID_1));
+    public static final BeaconId EXPECTED_ESTIMOTE_ID = new BeaconId(Utils.hexStringToByteArray(ESTIMOTE_ID));
+
+
+    public static final byte[] BYTES_FOR_BEACON_1 = hexStringToByteArray( ADVERTISEMENT_DATA_FLAGS + IBEACON_HEADER + BEACON_ID_1 + CALIBRATED_TX_VALUE );
+    public static final byte[] BYTES_FOR_SENSORBERG_BEACON_1 = hexStringToByteArray( ADVERTISEMENT_DATA_FLAGS + IBEACON_HEADER + BEACON_ID_1 + CALIBRATED_TX_VALUE );
+    public static final byte[] BYTES_FOR_ALIEN_BEACON_1 = hexStringToByteArray( ADVERTISEMENT_DATA_FLAGS + IBEACON_HEADER + ALIEN_ID_1 + CALIBRATED_TX_VALUE );
+
+
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/testUtils/TestBluetoothPlatform.java]: don't need the extra 2 spaces.

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/testUtils/TestBluetoothPlatform.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        fakeIBeaconSighting(BYTES_FOR_ALIEN_BEACON_1);
+    }
+
+    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+    public void fakeIBeaconSighting(byte[] bytesForFakeScan){
+        if (this.scanCallback != null){
+            this.scanCallback.onLeScan(null, -100, wrapWithZeroBytes(bytesForFakeScan, 62));
+        }
+    }
+
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/testUtils/TestBluetoothPlatform.java]: no space

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/testUtils/TestFileManager.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+    @Override
+    public File getFile(String fileName) {
+        try {
+            return File.createTempFile(fileName, null);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/testUtils/TestFileManager.java]: no space

File: android-sdk/src/androidTest/java/util/VolleyUtil.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+        public static RequestQueue getCachedVolleyQueue(Context context) {
+        Network network = new BasicNetwork(new OkHttpStack());
+        File cacheDir = new File(context.getCacheDir(), "volley-test-" + String.valueOf(System.currentTimeMillis()));
+        RequestQueue queue = new RequestQueue(new DiskBasedCache(cacheDir), network);
+        queue.start();
+
+        return queue;
+    }
+
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/util/VolleyUtil.java]: no space

File: android-sdk/src/main/java/com/sensorberg/BackgroundDetector.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+import com.sensorberg.sdk.internal.Platform;
+
+import android.annotation.TargetApi;
+import android.app.Activity;
+import android.app.Application;
+import android.os.Build;
+import android.os.Bundle;
+import android.os.Handler;
+
+
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/BackgroundDetector.java]: delete space

File: android-sdk/src/main/java/com/sensorberg/BackgroundDetector.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
+
+    }
+
+    @Override
+    public void onActivityDestroyed(Activity activity) {
+
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/BackgroundDetector.java]: delete spaces

File: android-sdk/src/main/java/com/sensorberg/SensorbergApplicationBootstrapper.java [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+        hostApplicationInForeground();
+    }
+
+    public void hostApplicationInBackground() {
+        Logger.log.applicationStateChanged("hostApplicationInBackground");
+        sendEmptyMessage(SensorbergService.MSG_APPLICATION_IN_BACKGROUND);
+        unRegisterFromPresentationDelegation();
+    }
+
+
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/SensorbergApplicationBootstrapper.java]: no space

File: android-sdk/src/main/java/com/sensorberg/di/Component.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    @NoArgsConstructor(access = AccessLevel.PRIVATE)
+    final class Initializer {
+
+        public static Component init(Application app) {
+            return DaggerComponent.builder()
+                    .providersModule(new ProvidersModule(app))
+                    .build();
+        }
+    }
+}
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/di/Component.java]: add a line,

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/TheInternalApplicationBootstrapperShould.java [hunk 18]
```diff
@@ -61,18 +84,23 @@ public void test_suppression_time() throws Exception {
...
-    public void test_send_only_once(){
+    public void test_send_only_once() {
         tested.onResolutionsFinished(Arrays.asList(beaconEventSentOnlyOnce));
         verify(tested, times(1)).presentBeaconEvent(any(BeaconEvent.class));
     }
+
+    public void test_should_return_the_sync_setting(){
+        Assertions.assertThat(tested.isSyncEnabled()).isTrue();
+    }
+
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/TheInternalApplicationBootstrapperShould.java]: no space

File: android-sdk/src/androidTest/java/com/sensorberg/sdk/testUtils/TestPlatformIdentifier.java [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    @Override
+    public void addDeviceInstallationIdentifierChangeListener(DeviceInstallationIdentifierChangeListener listener) {
+    }
+
+    @Override
+    public void addAdvertiserIdentifierChangeListener(AdvertiserIdentifierChangeListener listener) {
+    }
+
```
Comment (skraynick) [inline android-sdk/src/androidTest/java/com/sensorberg/sdk/testUtils/TestPlatformIdentifier.java]: no space

File: android-sdk/src/main/java/com/sensorberg/sdk/SensorbergService.java [hunk 20]
```diff
@@ -112,13 +163,18 @@ public static String stringFrom(int what) {
     }
 
     public static final String EXTRA_API_KEY = "com.sensorberg.android.sdk.intent.apiKey";
+
     public static final String EXTRA_BLUETOOTH_STATE = "com.sensorberg.android.sdk.intent.bluetoothState";
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/sdk/SensorbergService.java]: let's remove the extra space.

File: android-sdk/src/main/java/com/sensorberg/sdk/internal/AndroidBluetoothPlatform.java [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+                crashCallBackWrapper.setCallback(null);
+            }
+        }
+    }
+
+    @Override
+    public boolean isLeScanRunning() {
+        return leScanRunning;
+    }
+
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/sdk/internal/AndroidBluetoothPlatform.java]: no space

File: android-sdk/src/main/java/com/sensorberg/sdk/internal/AndroidHandlerManager.java [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+    @Override
+    public RunLoop getResolverRunLoop(RunLoop.MessageHandlerCallback callback) {
+        return new AndroidHandler(callback);
+    }
+
+    @Override
+    public RunLoop getBeaconPublisherRunLoop(RunLoop.MessageHandlerCallback callback) {
+        return new AndroidHandler(callback);
+    }
+
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/sdk/internal/AndroidHandlerManager.java]: no space

File: android-sdk/src/main/java/com/sensorberg/sdk/internal/AndroidPlatformIdentifier.java [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+        }
+    }
+
+    @TargetApi(Build.VERSION_CODES.DONUT)
+    private static String getAppLabel(Context application) {
+        PackageManager pm = application.getPackageManager();
+        ApplicationInfo ai = application.getApplicationInfo();
+        return String.valueOf(pm.getApplicationLabel(ai));
+    }
+
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/sdk/internal/AndroidPlatformIdentifier.java]: no space

File: android-sdk/src/main/java/com/sensorberg/sdk/internal/AndroidServiceScheduler.java [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+
+    @Override
+    public void scheduleRepeating(int MSG_type, long value, TimeUnit timeUnit) {
+        long millis = TimeUnit.MILLISECONDS.convert(value, timeUnit);
+        PendingIntent pendingIntent = getPendingIntent(MSG_type);
+        alarmManager.setInexactRepeating(AlarmManager.ELAPSED_REALTIME_WAKEUP, clock.elapsedRealtime() + millis, millis, pendingIntent);
+        repeatingPendingIntents.add(MSG_type);
+    }
+
+
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/sdk/internal/AndroidServiceScheduler.java]: no space

File: android-sdk/src/main/java/com/sensorberg/sdk/internal/AndroidServiceScheduler.java [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+    @Override
+    public void postToServiceDelayed(long delay, int type, Parcelable what, boolean surviveReboot) {
+        int index = postToServiceCounter.next();
+        postToServiceDelayed(delay, type, what, surviveReboot, index);
+    }
+
+
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/sdk/internal/AndroidServiceScheduler.java]: no space

File: android-sdk/src/main/java/com/sensorberg/sdk/internal/interfaces/HandlerManager.java [hunk 26]
```diff
@@ -0,0 +1,11 @@
+package com.sensorberg.sdk.internal.interfaces;
+
+public interface HandlerManager {
+
+    RunLoop getScannerRunLoop(RunLoop.MessageHandlerCallback callback);
+
+    RunLoop getResolverRunLoop(RunLoop.MessageHandlerCallback callback);
+
+    RunLoop getBeaconPublisherRunLoop(RunLoop.MessageHandlerCallback callback);
+
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/sdk/internal/interfaces/HandlerManager.java]: no space

File: android-sdk/src/main/java/com/sensorberg/sdk/settings/DefaultSettings.java [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+    public static final long DEFAULT_CLEAN_BEACONMAP_ON_RESTART_TIMEOUT = Constants.Time.ONE_MINUTE;
+
+    public static final long DEFAULT_MESSAGE_DELAY_WINDOW_LENGTH = Constants.Time.ONE_SECOND * 10;
+
+    static final long DEFAULT_MILLIS_BEETWEEN_RETRIES = 5 * Constants.Time.ONE_SECOND;
+
+    public static final long DEFAULT_CACHE_TTL = 30 * Constants.Time.ONE_DAY;
+
+    static final int DEFAULT_MAX_RETRIES = 3;
+
```
Comment (skraynick) [inline android-sdk/src/main/java/com/sensorberg/sdk/settings/DefaultSettings.java]: no space.

File: dev-app/src/main/java/com/sensorberg/sdk/demo/DemoActivity.java [hunk 28]
```diff
@@ -67,68 +72,95 @@ public long elapsedRealtime() {
...
+            }
+        }
+    }
+
+    public static Intent getIntent(Context context, Action action) {
+        Intent intent = new Intent(context, DemoActivity.class);
+        intent.putExtra(EXTRA_ACTION, action);
+        return intent;
+    }
+
```
Comment (skraynick) [inline dev-app/src/main/java/com/sensorberg/sdk/demo/DemoActivity.java]: no space

**Comment by skraynick** [comment]
@mrsasha I tried to run the tests. Sugar tests are all failing. I checked your pull, it appears that the sugar code didn't make it completely into your branch.  For ex. the db test helper class is no longer used so the db tables are not being created. Please check.

**Comment by mrsasha** [comment]
@skraynick can you check the tests now? I have a somewhat conflicting report - it works if I call them locally, but travis still complains for some reason.

**Comment by mrsasha** [comment]
@skraynick everything seems ok now.

**Pull request merged**